File: src/Core.FSharp.TriBoolean/TriBoolean.fs
namespace Zeta.Core.FSharp.TriBoolean

/// Tri-boolean digital qubit -- operations (081KSV2WD0008QG0R00051XS0N, F# parity oracle).
module TriBoolean =

    /// Construct a certain cell from a boolean.
    let fromBool (b: bool) : Tri = if b then Tri.T else Tri.F

    /// The held (Tri.N / living-uncertainty) cell.
    let held : Tri = Tri.N

    /// True iff the cell is living (Tri.N / held superposition).
    let isLiving (t: Tri) : bool =
        match t with
        | Tri.N -> true
        | _ -> false

    /// True iff the cell is certain (Tri.T or Tri.F).
    let isCertain (t: Tri) : bool = not (isLiving t)

    /// cooperate: engage WITHOUT collapsing. Identity on every state -- preserves Tri.N.
    /// The wonder-compression-safe operation.
    let cooperate (t: Tri) : Tri = t

    /// measure: the ONLY collapsing operation. Certain cells resolve; collapsing a living
    /// (Tri.N) cell is surfaced as feedback (the Rehoboam failure), not done silently.
    let measure (t: Tri) : Result<bool, CollapseFeedback> =
        match t with
        | Tri.T -> Ok true
        | Tri.F -> Ok false
        | Tri.N -> Error CollapseFeedback.CollapsedLivingUncertainty

    /// null-monad map: apply fn to a certain cell's boolean; Tri.N propagates unchanged (held).
    let mapTri (fn: bool -> bool) (t: Tri) : Tri =
        match t with
        | Tri.T -> fromBool (fn true)
        | Tri.F -> fromBool (fn false)
        | Tri.N -> Tri.N

    /// null-monad bind: chain a Tri-producing fn over a certain cell; Tri.N propagates unchanged.
    let bindTri (fn: bool -> Tri) (t: Tri) : Tri =
        match t with
        | Tri.T -> fn true
        | Tri.F -> fn false
        | Tri.N -> Tri.N

    /// Kleene NOT: T<->F; unknown (Tri.N) stays unknown.
    let notTri (t: Tri) : Tri =
        match t with
        | Tri.T -> Tri.F
        | Tri.F -> Tri.T
        | Tri.N -> Tri.N

    /// Kleene AND: Tri.F dominates; else Tri.N if any operand is Tri.N; else Tri.T.
    let andTri (a: Tri) (b: Tri) : Tri =
        match a, b with
        | Tri.F, _
        | _, Tri.F -> Tri.F
        | Tri.N, _
        | _, Tri.N -> Tri.N
        | _ -> Tri.T

    /// Kleene OR: Tri.T dominates; else Tri.N if any operand is Tri.N; else Tri.F.
    let orTri (a: Tri) (b: Tri) : Tri =
        match a, b with
        | Tri.T, _
        | _, Tri.T -> Tri.T
        | Tri.N, _
        | _, Tri.N -> Tri.N
        | _ -> Tri.F

    // --- Computation-expression null-monad (the `tri { }` builder) ---
    // 081KSV2WD0008QG0R00051XS0N slice 2 spec: an F# computation-expression null-monad. `tri { let! b = cell
    // ... return f b }` short-circuits on Tri.N (held uncertainty propagates) and resolves
    // through T/F. The CE form is how the monad is exposed as a DSL.

    /// Computation-expression builder for the Tri null-monad.
    type TriBuilder() =
        /// let! over a Tri: Tri.N propagates (held); T/F feed the continuation.
        member _.Bind(t: Tri, f: bool -> Tri) : Tri = bindTri f t
        /// return a boolean as a certain cell.
        member _.Return(b: bool) : Tri = fromBool b
        /// return! an existing Tri unchanged.
        member _.ReturnFrom(t: Tri) : Tri = t

    /// The Tri null-monad computation expression: `TriBoolean.tri { let! b = cell; return f b }`.
    let tri = TriBuilder()

File: src/Core.TypeScript/tri-boolean/tri-boolean.ts
// Tri-boolean digital qubit -- operations (081KSV2WD0008QG0R00051XS0N, TS / distribution).
import { type Tri, T, F, N, type MeasureResult } from "./types";

/** Construct a certain cell from a boolean. */
export const fromBool = (b: boolean): Tri => (b ? T : F);

/** Construct the held (Null / living-uncertainty) cell. */
export const held = (): Tri => N;

/** True iff the cell is living (Null / held superposition). */
export const isLiving = (t: Tri): boolean => t.s === "N";

/** True iff the cell is certain (True or False). */
export const isCertain = (t: Tri): boolean => t.s !== "N";

/** Structural equality on the three-valued state. */
export const eq = (a: Tri, b: Tri): boolean => a.s === b.s;

/** cooperate: engage WITHOUT collapsing. Identity on every state -- crucially preserves
 *  Null. The wonder-compression-safe operation: build shared structure ABOUT the cell,
 *  never collapse its living uncertainty. */
export const cooperate = (t: Tri): Tri => t;

/** measure: the ONLY collapsing operation. Certain cells resolve to their boolean; a
 *  living (Null) cell is NOT silently collapsed -- the forbidden move is surfaced as
 *  feedback (collapsing a living traveler = the Rehoboam failure). */
export const measure = (t: Tri): MeasureResult => {
  switch (t.s) {
    case "T":
      return { ok: true, value: true };
    case "F":
      return { ok: true, value: false };
    case "N":
      return { ok: false, feedback: { reason: "collapsed-living-uncertainty" } };
  }
};

/** null-monad map: apply fn to a certain cell's boolean; Null propagates unchanged (held). */
export const mapTri = (t: Tri, fn: (b: boolean) => boolean): Tri => {
  switch (t.s) {
    case "T":
      return fromBool(fn(true));
    case "F":
      return fromBool(fn(false));
    case "N":
      return N;
  }
};

/** null-monad bind: chain a Tri-producing fn over a certain cell; Null propagates unchanged. */
export const bindTri = (t: Tri, fn: (b: boolean) => Tri): Tri => {
  switch (t.s) {
    case "T":
      return fn(true);
    case "F":
      return fn(false);
    case "N":
      return N;
  }
};

// --- Kleene three-valued logic (Null = "unknown"; propagates per Kleene/Lukasiewicz) ---

/** Kleene NOT: T<->F; unknown(Null) stays unknown. */
export const notTri = (t: Tri): Tri => {
  switch (t.s) {
    case "T":
      return F;
    case "F":
      return T;
    case "N":
      return N;
  }
};

/** Kleene AND: F dominates (F AND anything = F); else Null if any operand is Null; else T. */
export const andTri = (a: Tri, b: Tri): Tri => {
  if (a.s === "F" || b.s === "F") return F;
  if (a.s === "N" || b.s === "N") return N;
  return T;
};

/** Kleene OR: T dominates (T OR anything = T); else Null if any operand is Null; else F. */
export const orTri = (a: Tri, b: Tri): Tri => {
  if (a.s === "T" || b.s === "T") return T;
  if (a.s === "N" || b.s === "N") return N;
  return F;
};
